libfwupdplugin/fu-block-partition.c

```diff
@@ -0,0 +1,260 @@
+
+	/* not changed */
+	if (g_strcmp0(priv->fs_uuid, fs_uuid) == 0)
+		return;
+
+	g_free(priv->fs_uuid);
+	priv->fs_uuid = fu_strsafe(fs_uuid, fs_uuidlen);
+}
+#endif
+
+static void
+fu_block_partition_incorporate(FuDevice *self, FuDevice *donor)
+{
+	FuBlockPartition *uself = FU_BLOCK_PARTITION(self);
+	FuBlockPartition *udonor = FU_BLOCK_PARTITION(donor);
+	FuBlockPartitionPrivate *priv = GET_PRIVATE(uself);
+
+	g_return_if_fail(FU_IS_BLOCK_PARTITION(self));
+	g_return_if_fail(FU_IS_BLOCK_PARTITION(donor));
+
+	if (priv->fs_type == NULL)
+		priv->fs_type = g_strdup(fu_block_partition_get_fs_type(udonor));
+	if (priv->fs_uuid == NULL)
+		priv->fs_uuid = g_strdup(fu_block_partition_get_fs_uuid(udonor));
+}
+
+/**
+ * fu_block_partition_get_fs_type:
+ * @self: a #FuBlockPartition
+ *
+ * Returns the filesystem type, e.g. `msdos`.
+ *
+ * Returns: string, or %NULL for unset
+ *
+ * Since: 2.0.2
+ **/
+const gchar *
+fu_block_partition_get_fs_type(FuBlockPartition *self)
+{
+	FuBlockPartitionPrivate *priv = GET_PRIVATE(self);
+	g_return_val_if_fail(FU_IS_BLOCK_PARTITION(self), NULL);
+	return priv->fs_type;
+}
+
+/**
+ * fu_block_partition_get_fs_uuid:
+ * @self: a #FuBlockPartition
+ *
+ * Returns the filesystem UUID.
+ *
+ * Returns: string, or %NULL for unset
+ *
+ * Since: 2.0.2
+ **/
+const gchar *
+fu_block_partition_get_fs_uuid(FuBlockPartition *self)
+{
+	FuBlockPartitionPrivate *priv = GET_PRIVATE(self);
+	g_return_val_if_fail(FU_IS_BLOCK_PARTITION(self), NULL);
+	return priv->fs_uuid;
+}
+
+static gboolean
+fu_block_partition_setup(FuDevice *device, GError **error)
+{
+	FuBlockPartition *self = FU_BLOCK_PARTITION(device);
+	FuBlockPartitionPrivate *priv = GET_PRIVATE(self);
+	FuDeviceEvent *event = NULL;
+	FuIOChannel *io_channel = fu_udev_device_get_io_channel(FU_UDEV_DEVICE(device));
+	g_autofree gchar *event_id = NULL;
+#ifdef HAVE_BLKID
+	gint rc;
+	const gchar *data;
+	gsize datalen = 0;
+	g_auto(blkid_probe) pr = NULL;
+#endif
+
+	/* need event ID */
+	if (fu_device_has_flag(FU_DEVICE(self), FWUPD_DEVICE_FLAG_EMULATED) ||
+	    fu_context_has_flag(fu_device_get_context(FU_DEVICE(self)),
+				FU_CONTEXT_FLAG_SAVE_EVENTS)) {
+		event_id = g_strdup_printf("FuBlockPartitionSetup:DeviceFile=%s",
+					   fu_udev_device_get_device_file(FU_UDEV_DEVICE(self)));
+	}
+
+	/* emulated */
+	if (fu_device_has_flag(FU_DEVICE(self), FWUPD_DEVICE_FLAG_EMULATED)) {
+		event = fu_device_load_event(FU_DEVICE(self), event_id, error);
+		if (event == NULL)
+			return FALSE;
+		priv->fs_type = g_strdup(fu_device_event_get_str(event, "FsType", NULL));
+		priv->fs_uuid = g_strdup(fu_device_event_get_str(event, "FsUuid", NULL));
+		return TRUE;
+	}
+
+	/* save */
+	if (event_id != NULL)
+		event = fu_device_save_event(FU_DEVICE(self), event_id);
+
+	/* sanity check */
+	if (io_channel == NULL) {
+		g_set_error_literal(error, FWUPD_ERROR, FWUPD_ERROR_NOT_SUPPORTED, "no device");
+		return FALSE;
+	}
+
+#ifdef HAVE_BLKID
+	/* probe */
+	pr = blkid_new_probe();
```

Comment:
does it set `errno`?

Reply:
Nope, it's only really going to fail for out-of-memory -- which is fatal for GLib anyway...